File: HyunWooLee/week9 - backtracking/2023-11-12-79_word_search.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """

        def dfs(row, col, curr_len):
            if row < 0 or len(board) <= row or col < 0 or len(board[0]) <= col:
                return False

            curr_word = word[len(word) - curr_len]

            if board[row][col] != curr_word:
                return False

            if curr_len == 1:
                return True

            prev = board[row][col]

            board[row][col] = None

            top = dfs(row - 1, col, curr_len - 1)
            down = dfs(row + 1, col, curr_len - 1)
            left = dfs(row, col - 1, curr_len - 1)
            right = dfs(row, col + 1, curr_len - 1)

            board[row][col] = prev

            return top or down or left or right

        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(i, j, len(word)):
                    return True

        return False

File: YonaKim/Week9/2023-11-12-79-Word-Search.java
class Solution {
  private char[][] board;
  private int numRows;
  private int numCols;

  public boolean searchWord(char[][] gameBoard, String targetWord) {
    this.board = gameBoard;
    this.numRows = gameBoard.length;
    this.numCols = gameBoard[0].length;

    for (int row = 0; row < numRows; ++row)
      for (int col = 0; col < numCols; ++col)
        if (backtrackSearch(row, col, targetWord, 0))
          return true;
    return false;
  }
  
  protected boolean backtrackSearch(int row, int col, String word, int index) {
    // Base case of recursion. If the word to be matched is empty, return true.
    if (index >= word.length())
      return true;

    // Check if the current state is invalid:
    // If the letter in the current cell does not match the first letter of the word,
    // or if the position of the cell is out of the boundary of the board.
    if (row < 0 || row == numRows || col < 0 || col == numCols
        || board[row][col] != word.charAt(index))
      return false;

    /* Step 3). Explore the neighbors in DFS */
    // Explore neighbors in DFS (backtracking)
    // Mark the current cell as visited (using any non-alphabetic letter)
    // Then iterate through the up, down, right, left directions (order of direction does not matter)
    boolean result = false;
    // Mark the path before the next exploration
    board[row][col] = '#';

    int[] rowOffsets = {0, 1, 0, -1};
    int[] colOffsets = {1, 0, -1, 0};
    for (int direction = 0; direction < 4; ++direction) {
      result = backtrackSearch(row + rowOffsets[direction], col + colOffsets[direction], word, index + 1);
      if (result)
        break;
    }

    // Revert the cell back to its original state
    // Return the result of recursion
    board[row][col] = word.charAt(index);
    return result;
  }
}
